packages/app/src/hooks/useNodeBodyHeight.tsx

```diff
@@ -0,0 +1,59 @@
+import { type NodeId } from '@ironclad/rivet-core';
+import { useCallback, useEffect, useMemo, useRef } from 'react';
+import { useRecoilValue } from 'recoil';
+import { nodesState } from '../state/graph';
+
+export interface NodeHeightCache {
+  get: (nodeId: NodeId) => number | undefined;
+
+  has: (nodeId: NodeId) => boolean;
+
+  set: (nodeId: NodeId, height: number | undefined) => void;
+}
+
+/**
+ * A cache of node heights. This is used when a node is unmounted and moved to the dragging list, since the node's
+ * body needs to be re-rendered in order to get its height. This cache allows us to avoid flickering when the node
+ * is first rendered in the dragging list.
+ */
+export const useNodeHeightCache = (): NodeHeightCache => {
+  const nodes = useRecoilValue(nodesState);
+  const ref = useRef<Record<string, number | undefined>>({});
+
+  const set = useCallback((nodeId: NodeId, height: number | undefined) => {
+    ref.current[nodeId] = height;
+  }, []);
+
+  const get = useCallback((nodeId: NodeId) => {
+    return ref.current[nodeId];
+  }, []);
+
+  const has = useCallback((nodeId: NodeId) => {
+    return nodeId in ref.current;
+  }, []);
+
+  return useMemo(() => {
+    const pre = ref.current;
+    ref.current = nodes.reduce((acc, next) => {
+      acc[next.id] = pre[next.id];
+      return acc;
+    }, {} as Record<string, number | undefined>);
+    return { set, get, has } satisfies NodeHeightCache;
+  }, [nodes, set, get, has]);
```

Comment:
This will reduce for all nodes every time any node changes, not a huge deal but it can add up

Reply:
Agreed.

I've updated the hook to separate the `set,get,has` interface of the cache from the implementation of the cache. I also changed it so that the cache is cleaned after an interval of node changes (i.e. 100 changes) to help boost performance.

Here is the modified code:

```
  /**
   * This hook removes nodes from the cache that have been deleted. To improve performance, we only clean up
   * the cache for every X nodes that are deleted (cache is just numbers).
   */
  useEffect(() => {
    if (garbageCollectionCount.current++ % GARBAGE_COLLECTION_INTERVAL !== 0) {
      ref.current = nodes.reduce((acc, next) => {
        acc[next.id] = ref.current[next.id];
        return acc;
      }, {} as Record<string, number | undefined>);
    }
  }, [nodes]);

  return useMemo(() => {
    return { set, get, has } satisfies HeightCache;
  }, [set, get, has]);
```